google/cloud/aiplatform/utils/featurestore_utils.py

```diff
@@ -0,0 +1,188 @@
+# -*- coding: utf-8 -*-
+
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import re
+from typing import Dict, Optional
+
+RESOURCE_ID_PATTERN_REGEX = r"[a-z_][a-z0-9_]{0,59}"
+FEATURESTORE_NAME_PATTERN_REGEX = (
+    r"projects\/(?P<project>[\w-]+)"
+    r"\/locations\/(?P<location>[\w-]+)"
+    r"\/featurestores\/(?P<featurestore_id>" + RESOURCE_ID_PATTERN_REGEX + r")"
+)
+ENTITY_TYPE_NAME_PATTERN_REGEX = (
+    FEATURESTORE_NAME_PATTERN_REGEX
+    + r"\/entityTypes\/(?P<entity_type_id>"
+    + RESOURCE_ID_PATTERN_REGEX
+    + r")"
+)
+FEATURE_NAME_PATTERN_REGEX = (
+    ENTITY_TYPE_NAME_PATTERN_REGEX
+    + r"\/features\/(?P<feature_id>"
+    + RESOURCE_ID_PATTERN_REGEX
+    + r")"
+)
+
+
+def validate_id(resource_id: str) -> bool:
+    """Validates feature store resource ID pattern."""
+    return bool(re.compile(r"^" + RESOURCE_ID_PATTERN_REGEX + r"$").match(resource_id))
+
+
+def validate_featurestore_name(featurestore_name: str) -> Dict[str, str]:
```

Comment:
Preference to name these `parse` instead of `validate`.

Also, is additional RESOURCE_ID validation the main reason we are not relying on the GAPIC parse methods[1]?

1. https://github.com/googleapis/python-aiplatform/blob/main/google/cloud/aiplatform_v1/services/featurestore_service/client.py#L230

Reply:
Yes, I am still evaluating the restrictions on Feature Store's resource ids, since we are enabling resource creation/ingestion through user provided df. My gut feeling is we may need some validation in the SDK layer but I think I will have clearer idea when implementing the df related methods. 

I will use the GAPIC methods for now and revisit later if they are needed.